presto-hive/src/main/java/com/facebook/presto/hive/ParquetHiveRecordCursor.java

```diff
@@ -429,7 +432,7 @@ public ReadContext init(
             ImmutableList.Builder<parquet.schema.Type> fields = ImmutableList.builder();
             for (HiveColumnHandle column : columns) {
                 if (!column.isPartitionKey() && column.getHiveColumnIndex() < messageType.getFieldCount()) {
-                    fields.add(messageType.getType(column.getName()));
+                    fields.add(getParquetType(configuration, column, messageType));
```

Comment:
I'm not sure this fixes the whole problem.  Looking at the code I think if the Hive metastore has columns `name:varchar, age:bigint` and the file has `age:bigint, name:varchar`, the reader will only flip the types and not the order, so the Presto query will see `name:bigint, age:varchar`.  Did you test this type of column reordering?